llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp

```diff
@@ -55,15 +56,91 @@ void InterBlockScheduling::markEpilogueBlocks() {
   }
 }
 
+std::map<MachineBasicBlock *, LivePhysRegs> calculateLiveOuts(MachineFunction *MF) {
+  std::queue<MachineBasicBlock *> WorkList;
+  std::set<MachineBasicBlock *> WorkListSet;
+  std::map<MachineBasicBlock *, LivePhysRegs> LiveIns;
+  std::map<MachineBasicBlock *, LivePhysRegs> LiveOuts;
+  const TargetRegisterInfo *TRI = MF->getSubtarget().getRegisterInfo();
+  for (MachineBasicBlock *MBB : post_order(MF)) {
+    LiveIns[MBB].init(*TRI);
+    LiveOuts[MBB].init(*TRI);
+    WorkListSet.insert(MBB);
+    WorkList.push(MBB);
+  }
+  while (!WorkList.empty()) {
+    MachineBasicBlock *MBB = WorkList.front();
+    WorkList.pop();
+    WorkListSet.erase(MBB);
+    LivePhysRegs CurrentLive;
+    CurrentLive.init(*TRI);
+
+    // Calculates LiveOuts by iterating over each successor of the current MBB
+    // and adding each successor's LiveIns to the current MBB's LiveOuts.
+    // This ensures LiveOuts is the union of LiveIns of all successor MBBs.
+    for (MachineBasicBlock *Succ : MBB->successors()) {
+      for (MCPhysReg Reg : LiveIns[Succ]) {
+        LiveOuts[MBB].addReg(Reg);
+        CurrentLive.addReg(Reg);
+      }
+    }
+    // Calculates CurrentLive by bottom-up traversal the MBB
+    for (MachineInstr &MI : llvm::reverse(*MBB)) {
+      CurrentLive.stepBackward(MI);
+    }
+    // Checks for convergence of LiveIns of MBB
+    bool Change = false;
```

Comment:
I would suggest moving the comparison logic for LiveRegs to a separate function. E.g.
```
bool liveRegsAreEqual(const LivePhysRegs &Lhs, const LivePhysRegs &Rhs) {
  return std::equal(Lhs.begin(), Lhs.end(), Rhs.begin(), Rhs.end());
}
```

That could even become an `operator==` for `LivePhysRegs`. With that function, I think we can make the code more concise, e.g.
```
    if (!liveRegsAreEqual(CurrentLive, LiveIns[MBB]) {
      // The liveins for MBB have changed, trigger a recomputation for MBB's preds
      for (MachineBasicBlock *Pred : MBB->predecessors()) {
        if (WorkListSet.find(Pred) == WorkListSet.end()) {
          WorkListSet.insert(Pred);
          WorkList.push(Pred);
        }
      }
    }
```